addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/socket/XiaomiSocket.java

```diff
@@ -0,0 +1,203 @@
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.mihome.internal.socket;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+/**
+ * Takes care of the communication with MiHome devices.
+ *
+ *
+ * @author Patrick Boos - Initial contribution
+ * @author Dimalo - JavaDoc and clean code
+ *
+ */
+public abstract class XiaomiSocket {
+
+    static final String MCAST_ADDR = "224.0.0.50";
+    private static final int BUFFER_LENGTH = 1024;
+    private static DatagramPacket datagramPacket = new DatagramPacket(new byte[BUFFER_LENGTH], BUFFER_LENGTH);
+
+    protected Thread socketReceiveThread;
+    private static List<XiaomiSocketListener> listeners = new CopyOnWriteArrayList<>();
+    private static final JsonParser PARSER = new JsonParser();
+
+    Logger logger = LoggerFactory.getLogger(this.getClass().getName());
+
+    static ArrayList<XiaomiSocket> openSockets = new ArrayList<XiaomiSocket>();
+
+    int port = 0;
+    DatagramSocket socket;
+
+    /**
+     * Sets up an {@link XiaomiSocket} with the MiHome multicast address and a random port
+     *
+     */
+    public XiaomiSocket() {
+        this(0);
+    }
+
+    /**
+     * Sets up an {@link XiaomiSocket} with the MiHome multicast address and a specific port
+     *
+     * @param port - the socket will be bound to this port
+     */
+    public XiaomiSocket(int port) {
+        this.port = port;
+
+        setupSocket();
+    }
+
+    abstract void setupSocket();
+
+    /**
+     * Interrupts the {@link ReceiverThread} and closes the {@link XiaomiSocket}.
+     */
+    private void closeSocket() {
+        synchronized (XiaomiSocket.class) {
+            if (socketReceiveThread != null) {
+                socketReceiveThread.interrupt();
+            }
+            if (socket != null) {
+                socket.close();
+                logger.debug("Socket closed");
+                socket = null;
+            }
+        }
+    }
+
+    /**
+     * Registers a {@link XiaomiSocketListener} to be called back, when data is received.
+     * If no {@link XiaomiSocket} exists, when the method is called, it is being set up.
+     *
+     * @param listener - {@link XiaomiSocketListener} to be called back
+     */
+    public synchronized void registerListener(XiaomiSocketListener listener) {
+        if (socket == null) {
+            setupSocket();
+        }
+        if (!listeners.contains(listener)) {
+            logger.trace("Adding socket listener {}", listener.toString());
```

Comment:
You don't have to call toString it will be called automatically, but only when the logging-level actually applies.